lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -880,9 +878,24 @@ bool HandshakeBuilder::visitHandshake(MuxOp op) {
   // Connect that to the select ready.
   rewriter.create<ConnectOp>(insertLoc, selectReady, resultValidAndReady);
 
+  // Since addresses coming from Handshake are IndexType and have a hardcoded
+  // 64-bit width in this pass, we may need to truncate down to the actual
+  // width used to index into the decoder.
+  uint64_t decodeWidth = argData.size();
+  size_t bitsNeeded = getNumIndexBits(decodeWidth);
+  size_t selectBits =
+      selectData.getType().cast<FIRRTLType>().getBitWidthOrSentinel();
+
+  if (selectBits > bitsNeeded) {
+    auto tailAmount = selectBits - bitsNeeded;
+    auto tailType = UIntType::get(op.getContext(), bitsNeeded);
+    selectData = rewriter.create<TailPrimOp>(insertLoc, tailType, selectData,
+                                             tailAmount);
+  }
+
   // Create a decoder for the select data.
   auto decodedSelect =
-      createDecoder(selectData, argData.size(), insertLoc, rewriter);
+      createDecoder(selectData, decodeWidth, insertLoc, rewriter);
```

Comment:
It seems we no longer need to pass in `decodeWidth` as the pad operation has been removed.

Reply:
Good point! I will remove. In a previous revision, I was stil using this width to compute the result type of the `dshl` operation, but then I realized that computation is exactly what `DShlPrimOp::getResultType` already does.